pumpp/task/key.py

```diff
@@ -0,0 +1,438 @@
+                mode_transpose_int = -1 * QUALITY[quality]
+                    
+            mode_profile_in_c = np.roll(c_major_profile, mode_transpose_int)
+            # Add the leading tone to the minor profiles
+            if quality == 'minor':
+                mode_profile_in_c[11] = 1
+            
+            # Now roll the profile again to get the right tonic.
+            pitch_profile = np.roll(mode_profile_in_c, tonic)
+
+        if not self.sparse:
+            tonic_vec = np.zeros(13, dtype=np.bool)
+            tonic_vec[tonic] = 1
+            tonic = tonic_vec
+
+        return (pitch_profile, tonic)
+
+    def empty(self, duration):
+        '''Empty key annotation
+
+        Parameters
+        ----------
+        duration : number
+            The length (in seconds) of the empty annotation
+
+        Returns
+        -------
+        ann : jams.Annotation
+            A key_mode annotation consisting of a single `no-key` observation.
+        '''
+        ann = super(KeyTransformer, self).empty(duration)
+
+        ann.append(time=0,
+                   duration=duration,
+                   value='N', confidence=0)
+
+        return ann
+
+    def transform_annotation(self, ann, duration):
+        '''Apply the key transformation.
+
+        Parameters
+        ----------
+        ann : jams.Annotation
+            The key_mode annotation
+
+        duration : number > 0
+            The target duration
+
+        Returns
+        -------
+        data : dict
+            data['pitch_profile'] : np.ndarray, shape=(n, 12)
+            data['tonic'] : np.ndarray, shape=(n, 13) or (n, 1)
+
+            `pitch_profile` is a binary matrix indicating pitch class
+            activation at each frame.
+
+            `tonic` is a one-hot matrix indicating the tonal center's 
+            pitch class at each frame.
+
+            If sparsely encoded, `tonic` is a integer
+            in the range [0, 12] where 12 indicates atonal.
+
+            If densely encoded, `tonic` has an extra
+            final dimension which is active when it is atonal.
+        '''
+        # get list of observations
+        intervals, keys = ann.to_interval_values()
+
+        # Get the dtype for tonic
+        if self.sparse:
+            dtype = np.int
+        else:
+            dtype = np.bool
+
+        # If we don't have any labeled intervals, fill in a 'N'
+        if not keys:
+            intervals = np.asarray([[0, duration]])
+            keys = ['N']
+
+        # Suppress all intervals not in the encoder
+        pitch_profiles = []
+        tonics = []
+
+        # default value when data is missing
+        if self.sparse:
+            fill = 12
+        else:
+            fill = False
+
+        for key in keys:
+            pitch_profile, tonic = self._encode_key_str(key)
+            pitch_profiles.append(pitch_profile)
+            tonics.append(tonic if type(tonic) is np.ndarray else [tonic])
```

Comment:
`type(...) is ...` is not robust to subclassing.  A better way to do this is `isinstance(tonic, np.ndarray)`

Reply:
👍